src/resourcemonitor/DefaultResourceMonitor.ts

```diff
@@ -0,0 +1,54 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import Logger from '../logger/Logger';
+import ResourceMonitor from '../resourcemonitor/ResourceMonitor';
+import AsyncScheduler from '../scheduler/AsyncScheduler';
+import IntervalScheduler from '../scheduler/IntervalScheduler';
+
+/** @internal */
+export default class DefaultResourceMonitor implements ResourceMonitor {
+  private readonly intervalScheduler: IntervalScheduler;
+
+  private count: number = 0;
+  private timestampMs: number = 0;
+
+  constructor(
+    private readonly logger: Logger,
+    private readonly dataPoints: number = 3,
+    private readonly intervalMs: number = 5000,
+    private readonly thresholdMs: number = 100,
+    private readonly waitTimeMs: number = 30000
+  ) {
+    this.intervalScheduler = new IntervalScheduler(this.intervalMs);
+  }
+
+  start(): void {
+    this.intervalScheduler.start(() => {
+      if (document.visibilityState !== 'visible') {
+        return;
+      }
+
+      const start = performance.now();
+      new AsyncScheduler().start(() => {
```

Comment:
Can we put some comment about how the logic work here?  It would be nice to have a summary instead of going through code :D